cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_ctrl_packages.go

```diff
@@ -428,3 +428,84 @@ func (s *Server) GetInstalledPackageDetail(ctx context.Context, request *corev1.
 	}
 	return response, nil
 }
+
+// CreateInstalledPackage creates an installed package managed by the 'kapp_controller' plugin
+func (s *Server) CreateInstalledPackage(ctx context.Context, request *corev1.CreateInstalledPackageRequest) (*corev1.CreateInstalledPackageResponse, error) {
+	log.Infof("+kapp-controller CreateInstalledPackage")
+
+	// Validate the request
+	if request.GetAvailablePackageRef().GetContext().GetNamespace() == "" || request.GetAvailablePackageRef().GetIdentifier() == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "required context or identifier not provided")
+	}
+	if request == nil || request.GetAvailablePackageRef() == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "no request AvailablePackageRef provided")
+	}
+	if request.GetName() == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "no request Name provided")
+	}
+	if request.GetTargetContext() == nil || request.GetTargetContext().GetNamespace() == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "no request TargetContext namespace provided")
+	}
+
+	// Retrieve the proper parameters from the request
+	packageRef := request.GetAvailablePackageRef()
+	cluster := packageRef.GetContext().GetCluster()
+	reconciliationOptions := request.GetReconciliationOptions()
+	pkgVersion := request.GetPkgVersionReference().GetVersion()
+	installedPackageName := request.GetName()
+	values := request.GetValues()
+	targetNamespace := request.GetTargetContext().GetNamespace()
+
+	if cluster == "" {
+		cluster = s.globalPackagingCluster
+	}
+
+	typedClient, _, err := s.GetClients(ctx, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	// fetch the package metadata
+	pkgMetadata, err := s.getPkgMetadata(ctx, cluster, targetNamespace, packageRef.Identifier)
+	if err != nil {
+		return nil, errorByStatus("get", "PackageMetadata", packageRef.Identifier, err)
+	}
+
+	// build a new secret object with the values
+	secret, err := s.newSecret(installedPackageName, values, targetNamespace)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, fmt.Sprintf("unable to create the Secret: %v", err))
+	}
+
+	// create the Secret in the cluster
+	createdSecret, err := typedClient.CoreV1().Secrets(targetNamespace).Create(ctx, secret, metav1.CreateOptions{})
```

Comment:
Hmm, I wonder if we should do this later in the process before `s.createPkgInstall` (or even after it, depending on how carvel handles that situation - I assume it'll see the secret as soon as created but we'd have to check). Main thing being to minimize the chance that we create the secret and leave it dangling if the other parts fail.

Actually, it could be simpler that we also clean up the secret if any other error occurs. I'd think we'd need to do this anyway otherwise the next creation would fail (unless you're avoiding collisions on the secret name). Even if not, we still should clean up imo.

Reply:
Good point! I'm moving the secret creation right before the `s.createPkgInstall` + adding a TODO to check if we really need to have the secret in advance (I'd say it is possible, but perhaps we will have to wait for an additional reconciliation cycle, but not sure).